src/backend/optimizer/plan/planner.c

```diff
@@ -1866,6 +1907,377 @@ grouping_planner(PlannerInfo *root, double tuple_fraction)
+						ListCell *lc;
+						Expr 	*expr;
+
+						/*
+						 * AQUMV
+						 * Process mv_query's tlist to pure-Var and no pure-Var expressions.
+						 * See details in README.cbdb.aqumv
+						 */
+						int i = 0;
+						foreach(lc, mvQuery->targetList)
+						{
+							i++;
+							TargetEntry* tle = lfirst_node(TargetEntry, lc);
+							expr = tle->expr;
+							if(IsA(expr, Var))
+							{
+								mv_pure_vars = lappend(mv_pure_vars, expr);
+								mv_pure_vars_index = lappend_int(mv_pure_vars_index, i);
+							}
+							else
+								mv_nonpure_vars_index = lappend_int(mv_nonpure_vars_index, i);
+						}
+
+						context->mv_pure_vars = mv_pure_vars;
+						context->mv_pure_vars_index = mv_pure_vars_index;
+						context->mv_nonpure_vars_index = mv_nonpure_vars_index;
+						context->mv_tlist = mvRelQueryTree->targetList;
+						context->mv_query_tlist = mvQuery->targetList;
+						context->has_unmatched = false;
+
+						/* Sort nonpure vars expression, prepare for Greedy Algorithm. */
+						aqumv_process_nonpure_vars_expr(context);
+
+						List	*mv_final_tlist = NIL; /* Final target list we want to rewrite to. */
+
+						/*
+						 * Process and rewrite target list, return false if failed.
+						 */
+						if(!aqumv_process_targetlist(context, parse->targetList, &mv_final_tlist))
+							continue;
+
+						/*
+						 * AQUMV
+						 * Process all quals to conjunctive normal form.
+						 * 
+						 * We assume that the selection predicates of view and query expressions
+						 * have been converted into conjunctive normal form(CNF) before we process
+						 * them.
+						 * AQUMV_MVP: no having quals now.
+						 */
+						preprocess_qual_conditions(subroot, (Node *) mvQuery->jointree);
+
+						List *post_quals = NIL;
+						/*
+						 * Process quals, return false if failed. 
+						 * Else, post_quals are filled if there were. 
+						 * Like process target list, post_quals is used later to see if we could
+						 * rewrite and apply it to mv relation.
+						 */
+						if(!aqumv_process_from_quals(parse->jointree->quals, mvQuery->jointree->quals, &post_quals))
+							continue;
+
+						/* Rewrite post_quals, return false if failed. */
+						post_quals = (List *)aqumv_adjust_sub_matched_expr_mutator((Node *)post_quals, context);
+						if (context->has_unmatched)
+							continue;
+
+						/*
+						 * Here! We succeed to rewrite a new SQL.
+						 * Begin to replace all guts.
+						 */
+						mvQuery->targetList = mv_final_tlist;
+
+						/*
+						 * AQUMV
+						 * NB: Update processed_tlist again in case that tlist has been changed. 
+						 */
+						preprocess_targetlist(subroot);
+
+						/*
+						 * AQUMV
+						 * NB: Correct the final_locus as we select from another realtion now.
+						 */
+						PathTarget *newtarget = make_pathtarget_from_tlist(subroot->processed_tlist);
+						subroot->final_locus = cdbllize_get_final_locus(root, newtarget);
```

Comment:
cdbllize_get_final_locus(root, newtarget), why not cdbllize_get_final_locus(`subroot` , newtarget) ?

Reply:
You're right, fixed.